File: verification/scripts/r/validate_extended_datasets.R
# ============================================================================
# Extended Datasets Validation - R Reference Implementation
# ============================================================================
# This script validates the Rust OLS implementation against R using synthetic
# and real-world datasets designed to test edge cases and stress conditions.
#
# Output: verification/datasets/results/r/*.json
# ============================================================================

cat("==============================================================\n")
cat("Extended Datasets Validation - R Reference Implementation\n")
cat("==============================================================\n")

# Add user library to path (helps find packages installed via RGUI)
user_lib <- file.path(Sys.getenv("USERPROFILE"), "Documents", "R", "win-library", "4.4")
if (dir.exists(user_lib)) {
    .libPaths(c(user_lib, .libPaths()))
}

# Load required packages
required_packages <- c("lmtest", "car", "tseries")
for (pkg in required_packages) {
    if (!require(pkg, character.only = TRUE, quietly = TRUE)) {
        stop(sprintf("Package '%s' is required. Install with: install.packages('%s')", pkg, pkg))
    }
}

# Set up output directory (relative to script location)
script_dir <- dirname(sys.frame(1)$filename)
output_dir <- file.path(script_dir, "..", "..", "..", "results", "r")
dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)

# Set path to datasets (relative to script location)
datasets_dir <- file.path(script_dir, "..", "..", "..", "datasets", "csv")

# ============================================================================
# Helper Functions
# ============================================================================

format_vector <- function(vec) {
    if (is.null(vec) || length(vec) == 0) return("[]")
    return(sprintf("[%s]", paste(vec, collapse = ", ")))
}

run_regression <- function(dataset_name, df, y_var, x_vars = NULL) {
    cat(sprintf("\n=== %s ===\n", dataset_name))

    # Build formula
    if (is.null(x_vars)) {
        x_vars <- setdiff(names(df), y_var)
    }
    formula_str <- paste(y_var, "~", paste(x_vars, collapse = " + "))
    formula <- as.formula(formula_str)

    # Fit model (with error handling)
    model <- tryCatch(lm(formula, data = df), error = function(e) {
        cat(sprintf("  Error: %s\n", conditionMessage(e)))
        return(NULL)
    })

    if (is.null(model)) return(NULL)

    # Extract results
    n <- nrow(df)
    k <- length(x_vars)
    df_residual <- model$df.residual

    coefficients <- as.vector(coef(model))
    std_errors <- as.vector(summary(model)$coefficients[, "Std. Error"])
    t_stats <- as.vector(summary(model)$coefficients[, "t value"])
    p_values <- as.vector(summary(model)$coefficients[, "Pr(>|t|)"])

    r_squared <- summary(model)$r.squared
    adj_r_squared <- summary(model)$adj.r.squared

    # Handle cases where F-statistic might be NA (e.g., perfect fit)
    f_stat <- tryCatch(summary(model)$fstatistic, error = function(e) NULL)
    if (!is.null(f_stat)) {
        f_statistic <- f_stat[1]
        f_p_value <- pf(f_statistic[1], f_statistic[2], f_statistic[3], lower.tail = FALSE)
    } else {
        f_statistic <- NA
        f_p_value <- NA
    }

    # Confidence intervals
    ci <- confint(model, level = 0.95)
    ci_lower <- as.vector(ci[, 1])
    ci_upper <- as.vector(ci[, 2])

    # VIF for multiple regression
    vif_results <- NULL
    if (k >= 2) {
        vif_values <- tryCatch(car::vif(model), error = function(e) NULL)
        if (!is.null(vif_values)) {
            vif_results <- list(
                variables = names(vif_values),
                vif = as.numeric(vif_values)
            )
        }
    }

    # Output JSON
    json_output <- sprintf('{
  "dataset": "%s",
  "n": %d,
  "k": %d,
  "df_residual": %d,
  "coefficients": %s,
  "std_errors": %s,
  "t_stats": %s,
  "p_values": %s,
  "r_squared": %.15f,
  "adj_r_squared": %.15f,
  "f_statistic": %s,
  "f_p_value": %s,
  "ci_lower": %s,
  "ci_upper": %s,
  "vif": %s
}',
        dataset_name,
        n, k, df_residual,
        format_vector(coefficients),
        format_vector(std_errors),
        format_vector(t_stats),
        format_vector(p_values),
        r_squared, adj_r_squared,
        ifelse(is.na(f_statistic), "null", as.character(f_statistic)),
        ifelse(is.na(f_p_value), "null", as.character(f_p_value)),
        format_vector(ci_lower),
        format_vector(ci_upper),
        if (!is.null(vif_results)) {
            sprintf('{"variables": ["%s"], "vif": %s}',
                    paste(vif_results$variables, collapse = '", "'),
                    paste(vif_results$vif, collapse = ", "))
        } else {
            "null"
        }
    )

    # Write to file
    output_file <- file.path(output_dir, paste0(gsub(" ", "_", tolower(dataset_name)), ".json"))
    writeLines(json_output, output_file)
    cat(sprintf("  Wrote: %s\n", basename(output_file)))

    # Print summary
    cat(sprintf("    n = %d, k = %d\n", n, k))
    cat(sprintf("    R² = %.6f, Adj R² = %.6f\n", r_squared, adj_r_squared))
    cat(sprintf("    F(%d, %d) = %.4f, p = %.6f\n", k, df_residual, f_statistic, f_p_value))
    if (!is.null(vif_results)) {
        cat(sprintf("    VIF: %s\n", paste(round(vif_results$vif, 2), collapse = ", ")))
    }

    invisible(list(
        dataset = dataset_name,
        n = n,
        k = k,
        r_squared = r_squared,
        vif = if (!is.null(vif_results)) vif_results$vif else NULL
    ))
}

# ============================================================================
# Load and Test Datasets
# ============================================================================

# --- Synthetic Datasets ---

# 1. Simple Linear
synthetic_simple <- read.csv(file.path(datasets_dir, "synthetic_simple_linear.csv"))
run_regression("Synthetic Simple Linear", synthetic_simple, "y", c("x"))

# 2. Multiple Regression
synthetic_multiple <- read.csv(file.path(datasets_dir, "synthetic_multiple.csv"))
run_regression("Synthetic Multiple", synthetic_multiple, "y", c("x1", "x2", "x3"))

# 3. Collinear (should fail or show high VIF)
synthetic_collinear <- read.csv(file.path(datasets_dir, "synthetic_collinear.csv"))
cat("\n--- Testing Collinear Dataset (expecting singular matrix or extreme VIF) ---\n")
tryCatch({
    run_regression("Synthetic Collinear", synthetic_collinear, "y", c("x1", "x2", "x3"))
}, error = function(e) {
    cat(sprintf("  Expected error: %s\n", conditionMessage(e)))
})

# 4. Heteroscedastic
synthetic_hetero <- read.csv(file.path(datasets_dir, "synthetic_heteroscedastic.csv"))
run_regression("Synthetic Heteroscedastic", synthetic_hetero, "y", c("x"))

# 5. Nonlinear
synthetic_nonlinear <- read.csv(file.path(datasets_dir, "synthetic_nonlinear.csv"))
run_regression("Synthetic Nonlinear", synthetic_nonlinear, "y", c("x"))

# 6. Nonnormal
synthetic_nonnormal <- read.csv(file.path(datasets_dir, "synthetic_nonnormal.csv"))
run_regression("Synthetic Nonnormal", synthetic_nonnormal, "y", c("x"))

# 7. Autocorrelated
synthetic_auto <- read.csv(file.path(datasets_dir, "synthetic_autocorrelated.csv"))
run_regression("Synthetic Autocorrelated", synthetic_auto, "y", c("x"))

# 8. High VIF
synthetic_high_vif <- read.csv(file.path(datasets_dir, "synthetic_high_vif.csv"))
run_regression("Synthetic High VIF", synthetic_high_vif, "y", NULL)

# 9. Outliers
synthetic_outliers <- read.csv(file.path(datasets_dir, "synthetic_outliers.csv"))
run_regression("Synthetic Outliers", synthetic_outliers, "y", NULL)

# 10. Small Sample
synthetic_small <- read.csv(file.path(datasets_dir, "synthetic_small.csv"))
run_regression("Synthetic Small", synthetic_small, "y", NULL)

# --- Real-World Datasets ---

# Longley (famous multicollinearity test case)
longley <- read.csv(file.path(datasets_dir, "longley.csv"))
# Use Employed as dependent, other columns as predictors
longley_y <- "Employed"
longley_x <- setdiff(names(longley), c(longley_y, "Year", "Unnamed..0"))
run_regression("Longley", longley, longley_y, longley_x)

# Mtcars
mtcars <- read.csv(file.path(datasets_dir, "mtcars.csv"))
run_regression("Mtcars", mtcars, "mpg", c("cyl", "disp", "hp", "wt", "qsec"))

# Bodyfat
bodyfat <- read.csv(file.path(datasets_dir, "bodyfat.csv"))
run_regression("Bodyfat", bodyfat, names(bodyfat)[1], NULL)

# Prostate
prostate <- read.csv(file.path(datasets_dir, "prostate.csv"))
run_regression("Prostate", prostate, names(prostate)[ncol(prostate)], NULL)

cat("\n==============================================================\n")
cat("Extended validation complete!\n")
cat(sprintf("Results saved to: %s\n", output_dir))
cat("==============================================================\n")

File: verification/scripts/python/validate_extended_datasets.py
"""
Extended Datasets Validation - Python Reference Implementation
===========================================================

This script validates the Rust OLS implementation against Python (statsmodels)
using synthetic and real-world datasets designed to test edge cases and stress conditions.

Output: verification/datasets/results/python/*.json
"""

import os
import sys
import json
import numpy as np
import pandas as pd
import statsmodels.api as sm
from statsmodels.stats.outliers_influence import variance_inflation_factor

# Add parent directory to path for imports
script_dir = os.path.dirname(os.path.abspath(__file__))
sys.path.insert(0, script_dir)

# Set up output directory (relative to script location)
# Output: verification/results/python/
output_dir = os.path.join(script_dir, "..", "..", "..", "results", "python")
os.makedirs(output_dir, exist_ok=True)

# Set path to datasets (relative to script location)
datasets_dir = os.path.join(script_dir, "..", "..", "..", "datasets", "csv")

def run_regression(dataset_name, df, y_var, x_vars=None):
    """Run OLS regression and save results to JSON."""
    print(f"\n=== {dataset_name} ===")

    # Prepare data
    y = df[y_var]
    if x_vars is None:
        X = df.drop(columns=[y_var])
    else:
        X = df[x_vars]

    # Add intercept
    X = sm.add_constant(X)

    # Fit model
    model = sm.OLS(y, X).fit()

    # Extract results
    n = len(y)
    k = len(x_vars) if x_vars else len(X.columns) - 1
    df_residual = model.df_resid

    coefficients = model.params.tolist()
    std_errors = model.bse.tolist()
    t_stats = model.tvalues.tolist()
    p_values = model.pvalues.tolist()

    r_squared = model.rsquared
    adj_r_squared = model.rsquared_adj
    f_statistic = model.fvalue
    f_p_value = model.f_pvalue

    # Confidence intervals
    ci = model.conf_int(alpha=0.05)
    ci_lower = ci[:, 0].tolist()
    ci_upper = ci[:, 1].tolist()

    # VIF for multiple regression
    vif_results = None
    if k >= 2:
        vif_values = [variance_inflation_factor(X.values, i)
                      for i in range(1, len(X.columns))]
        vif_results = {
            "variables": list(X.columns[1:]),
            "vif": vif_values
        }

    # Output JSON
    json_output = json.dumps({
        "dataset": dataset_name,
        "n": n,
        "k": k,
        "df_residual": df_residual,
        "coefficients": coefficients,
        "std_errors": std_errors,
        "t_stats": t_stats,
        "p_values": p_values,
        "r_squared": r_squared,
        "adj_r_squared": adj_r_squared,
        "f_statistic": f_statistic,
        "f_p_value": f_p_value,
        "ci_lower": ci_lower,
        "ci_upper": ci_upper,
        "vif": vif_results
    }, indent=2)

    # Write to file
    safe_name = dataset_name.lower().replace(" ", "_")
    output_file = os.path.join(output_dir, f"{safe_name}.json")
    with open(output_file, "w") as f:
        f.write(json_output)
    print(f"  Wrote: {os.path.basename(output_file)}")

    # Print summary
    print(f"    n = {n}, k = {k}")
    print(f"    R² = {r_squared:.6f}, Adj R² = {adj_r_squared:.6f}")
    print(f"    F({k}, {df_residual}) = {f_statistic:.4f}, p = {f_p_value:.6f}")
    if vif_results:
        vif_str = ", ".join([f"{v:.2f}" for v in vif_results["vif"]])
        print(f"    VIF: {vif_str}")

    return {
        "dataset": dataset_name,
        "n": n,
        "k": k,
        "r_squared": r_squared,
        "vif": vif_results["vif"] if vif_results else None
    }


def main():
    """Run validation on all datasets."""
    print("=" * 60)
    print("Extended Datasets Validation - Python Reference")
    print("=" * 60)

    results = []

    # --- Synthetic Datasets ---

    # 1. Simple Linear
    synthetic_simple = pd.read_csv(os.path.join(datasets_dir, "synthetic_simple_linear.csv"))
    results.append(run_regression("Synthetic Simple Linear", synthetic_simple, "y", ["x"]))

    # 2. Multiple Regression
    synthetic_multiple = pd.read_csv(os.path.join(datasets_dir, "synthetic_multiple.csv"))
    results.append(run_regression("Synthetic Multiple", synthetic_multiple, "y", ["x1", "x2", "x3"]))

    # 3. Collinear
    synthetic_collinear = pd.read_csv(os.path.join(datasets_dir, "synthetic_collinear.csv"))
    print("\n--- Testing Collinear Dataset (expecting high VIF or singular matrix) ---")
    try:
        results.append(run_regression("Synthetic Collinear", synthetic_collinear, "y", ["x1", "x2", "x3"]))
    except np.linalg.LinAlgError as e:
        print(f"  Expected error: {e}")

    # 4. Heteroscedastic
    synthetic_hetero = pd.read_csv(os.path.join(datasets_dir, "synthetic_heteroscedastic.csv"))
    results.append(run_regression("Synthetic Heteroscedastic", synthetic_hetero, "y", ["x"]))

    # 5. Nonlinear
    synthetic_nonlinear = pd.read_csv(os.path.join(datasets_dir, "synthetic_nonlinear.csv"))
    results.append(run_regression("Synthetic Nonlinear", synthetic_nonlinear, "y", ["x"]))

    # 6. Nonnormal
    synthetic_nonnormal = pd.read_csv(os.path.join(datasets_dir, "synthetic_nonnormal.csv"))
    results.append(run_regression("Synthetic Nonnormal", synthetic_nonnormal, "y", ["x"]))

    # 7. Autocorrelated
    synthetic_auto = pd.read_csv(os.path.join(datasets_dir, "synthetic_autocorrelated.csv"))
    results.append(run_regression("Synthetic Autocorrelated", synthetic_auto, "y", ["x"]))

    # 8. High VIF
    synthetic_high_vif = pd.read_csv(os.path.join(datasets_dir, "synthetic_high_vif.csv"))
    results.append(run_regression("Synthetic High VIF", synthetic_high_vif, "y", None))

    # 9. Outliers
    synthetic_outliers = pd.read_csv(os.path.join(datasets_dir, "synthetic_outliers.csv"))
    results.append(run_regression("Synthetic Outliers", synthetic_outliers, "y", None))

    # 10. Small Sample
    synthetic_small = pd.read_csv(os.path.join(datasets_dir, "synthetic_small.csv"))
    results.append(run_regression("Synthetic Small", synthetic_small, "y", None))

    # --- Real-World Datasets ---

    # Longley
    longley = pd.read_csv(os.path.join(datasets_dir, "longley.csv"))
    longley_y = "Employed"
    longley_x = [col for col in longley.columns if col not in [longley_y, "Year", "Unnamed: 0"]]
    results.append(run_regression("Longley", longley, longley_y, longley_x))

    # Mtcars
    mtcars = pd.read_csv(os.path.join(datasets_dir, "mtcars.csv"))
    results.append(run_regression("Mtcars", mtcars, "mpg", ["cyl", "disp", "hp", "wt", "qsec"]))

    # Bodyfat
    bodyfat = pd.read_csv(os.path.join(datasets_dir, "bodyfat.csv"))
    bodyfat_y = bodyfat.columns[0]
    results.append(run_regression("Bodyfat", bodyfat, bodyfat_y, None))

    # Prostate
    prostate = pd.read_csv(os.path.join(datasets_dir, "prostate.csv"))
    prostate_y = prostate.columns[-1]
    results.append(run_regression("Prostate", prostate, prostate_y, None))

    print("\n" + "=" * 60)
    print("Extended validation complete!")
    print(f"Results saved to: {output_dir}")
    print("=" * 60)

    return results


if __name__ == "__main__":
    main()
